### PR TITLE
Deprecate date and date_upto field from Project model

### LINKED ISSUE #383: Remove Project's date, date_upto and datelocation fields
Project dates should be calculated from the schedule. We can offer cached fields to make queries faster, but these will be datetime fields, not date fields, so `date` and `date_upto` should be replaced with `schedule_start_at` and `schedule_end_at`.

Since the `date` and `date_upto` fields are used in UI for the schedule editor, this needs attention.

The `datelocation` field is also deprecated by the `location` field introduced in #287 and can be removed (data may need to be migrated).

### PR DESCRIPTION
**Pull request by vidya-ram**
Fixes #383.

### CODE REVIEW

File: funnel/forms/project.py [hunk 1]
```diff
@@ -34,18 +34,7 @@ class ProjectForm(forms.Form):
...
-        __("End date"),
-        validators=[
-            forms.validators.DataRequired(__("End date is required")),
-            forms.validators.GreaterThanEqualTo('date', __("End date cannot be before start date"))
-            ]
-        )
-    tagline = forms.StringField(__("Tagline"),
-        validators=[forms.validators.DataRequired(), forms.validators.Length(max=250)],
-        filters=[forms.filters.strip()],
+    tagline = forms.StringField(__("Tagline"), validators=[forms.validators.DataRequired(), forms.validators.Length(max=250)],
```
Comment (jace) [inline funnel/forms/project.py]: `tagline` needs the `strip` filter.

File: funnel/models/project.py [hunk 2]
```diff
@@ -189,30 +187,32 @@ def datelocation(self):
...
-            strf_date = "%d %b %Y"
-        datelocation = datelocation_format.format(
-            date=self.date.strftime(strf_date),
-            date_upto=self.date_upto.strftime("%d %b"),
-            year=self.date.year)
-        return datelocation if not self.location else u', '.join([datelocation, self.location])
+        daterange = u""
+        if self.schedule_start_at and self.schedule_end_at:
+            daterange_format = u"{schedule_start_at}–{schedule_end_at} {year}"
+            if self.schedule_start_at == self.schedule_end_at:
```
Comment (jace) [inline funnel/models/project.py]: This will not happen since they are datetime, not date. Compare the date values after casting to project timezone.

File: funnel/models/project.py [hunk 3]
```diff
@@ -189,30 +187,32 @@ def datelocation(self):
...
+                # if both dates are same, in case of single day project
+                strf_date = ""
+                daterange_format = u"{schedule_end_at} {year}"
```
Comment (jace) [inline funnel/models/project.py]: Since you are rendering dates instead of datetimes, change the parameter names. I read this code expecting the actual `schedule_end_at` was being rendered here and only after scrolling down realised it was the date portion. Use `start_date` and `end_date`.

File: funnel/models/project.py [hunk 4]
```diff
@@ -189,30 +187,32 @@ def datelocation(self):
...
+            elif self.schedule_start_at.month != self.schedule_end_at.month:
+                # If multi-day event across months
+                strf_date = "%d %b"
+            elif self.schedule_start_at.year != self.schedule_end_at.year:
+                # if the start date and end dates are in different years,
+                strf_date = "%d %b %Y"
+            daterange = daterange_format.format(
+                schedule_start_at=self.schedule_start_at.astimezone(self.timezone).strftime(strf_date),
+                schedule_end_at=self.schedule_end_at.astimezone(self.timezone).strftime("%d %b"),
+                year=self.schedule_start_at.astimezone(self.timezone).year)
```
Comment (jace) [inline funnel/models/project.py]: Year has to be from the end date.

File: funnel/templates/project.html.jinja2 [hunk 5]
```diff
@@ -14,7 +14,7 @@
     "@type": "Event",
     "name": "{{ project.title }}",
     "url": "{{ project.url_for(_external=true) }}",
-    "startDate": "{{ project.date }}",
+    "startDate": "{{ project.schedule_start_at }}",
```
Comment (jace) [inline funnel/templates/project.html.jinja2]: These need `.isoformat()` as far as I can tell. Check the spec.

File: funnel/templates/project.html.jinja2 [hunk 6]
```diff
@@ -35,7 +35,7 @@
      ],
     {% endif -%}
     "description": "{{ project.tagline|safe }}",
-    "endDate": "{{ project.date_upto }}",
+    "endDate": "{{ project.schedule_end_at }}",
```
Comment (jace) [inline funnel/templates/project.html.jinja2]: Likewise.

File: funnel/views/project.py [hunk 7]
```diff
@@ -28,8 +28,8 @@ def project_data(project):
         'title': project.title,
         'datelocation': project.datelocation,
         'timezone': project.timezone.zone,
-        'start': project.date.isoformat() if project.date else None,
-        'end': project.date_upto.isoformat() if project.date_upto else None,
+        'start': project.schedule_start_at.date().isoformat() if project.schedule_start_at else None,
+        'end': project.schedule_end_at.date().isoformat() if project.schedule_end_at else None,
```
Comment (jace) [inline funnel/views/project.py]: Cast into local timezone first! Use `project.schedule_start_at.astimezone(project.timezone).date()`

File: funnel/views/schedule.py [hunk 8]
```diff
@@ -145,7 +145,7 @@ class ProjectScheduleView(ProjectViewMixin, UrlForView, ModelView):
     def schedule(self):
         schedule_transition_form = ProjectScheduleTransitionForm(obj=self.obj)
         return dict(project=self.obj,
-            from_date=date_js(self.obj.date), to_date=date_js(self.obj.date_upto),
+            from_date=date_js(self.obj.schedule_start_at), to_date=date_js(self.obj.schedule_end_at),
```
Comment (jace) [inline funnel/views/schedule.py]: The data type is changing from date to datetime. Does this affect the template?

File: funnel/views/schedule.py [hunk 9]
```diff
@@ -193,13 +193,8 @@ def edit_schedule(self):
...
             'scheduled': session_list_data(self.obj.scheduled_sessions, with_modal_url='edit', with_delete_url=True)
             }
-        # Set the proper range for the calendar to allow for date changes
-        first_session = Session.query.filter(Session.scheduled, Session.project == self.obj).order_by(Session.start.asc()).first()
-        last_session = Session.query.filter(Session.scheduled, Session.project == self.obj).order_by(Session.end.desc()).first()
-        from_date = (first_session and first_session.start.date() < self.obj.date and first_session.start) or self.obj.date
-        to_date = (last_session and last_session.start.date() > self.obj.date_upto and last_session.start) or self.obj.date_upto
         return dict(project=self.obj, proposals=proposals,
-            from_date=date_js(from_date), to_date=date_js(to_date),
+            from_date=date_js(self.obj.schedule_start_at), to_date=date_js(self.obj.schedule_end_at),
```
Comment (jace) [inline funnel/views/schedule.py]: Likewise. Is this compatible?

File: funnel/views/session.py [hunk 10]
```diff
@@ -90,7 +90,7 @@ class SessionView(SessionViewMixin, UrlForView, ModelView):
     @render_with('schedule.html.jinja2', json=True)
     def view(self):
         return dict(project=self.obj.project, active_session=session_data(self.obj, with_modal_url='view_popup'),
-            from_date=date_js(self.obj.project.date), to_date=date_js(self.obj.project.date_upto),
+            from_date=date_js(self.obj.project.schedule_start_at), to_date=date_js(self.obj.project.schedule_end_at),
```
Comment (jace) [inline funnel/views/session.py]: Likewise again.

File: funnel/views/schedule.py (hunk 8 above)
Comment (iambibhas) [reply funnel/views/schedule.py]: It doesn't. Previous they were timestamps created from `date` object, now they are timestamps created from `datetime` object.

File: funnel/templates/project.html.jinja2 (hunk 5 above)
Comment (iambibhas) [reply funnel/templates/project.html.jinja2]: Using `isoformat()`.

File: funnel/templates/project.html.jinja2 (hunk 6 above)
Comment (iambibhas) [reply funnel/templates/project.html.jinja2]: Using `isoformat()`.

File: funnel/models/project.py [hunk 11]
```diff
@@ -210,30 +214,34 @@ def datelocation(self):
...
-        elif self.date.year != self.date_upto.year:
-            # if the start date and end dates are in different years,
-            strf_date = "%d %b %Y"
-        datelocation = datelocation_format.format(
-            date=self.date.strftime(strf_date),
-            date_upto=self.date_upto.strftime("%d %b"),
-            year=self.date.year)
-        return datelocation if not self.location else u', '.join([datelocation, self.location])
+        daterange = u""
+        if self.schedule_start_at and self.schedule_end_at:
```
Comment (jace) [inline funnel/models/project.py]: Don't need to test for both, since sessions already enforce that `start_at` and `end_at` must exist together. If there is a `schedule_start_at`, then `schedule_end_at` is also guaranteed to exist.

Comment (jace) [reply funnel/models/project.py]: However, the test is harmless as both properties are cached. **But** you need to test with `is not None` because `date(0, 0, 0)` is falsy.

File: funnel/models/project.py [hunk 12]
```diff
@@ -549,16 +557,14 @@ def roles_for(self, actor=None, anchors=()):
...
+            Project.state.PUBLISHED))  # NOQA
 
 
 Profile.draft_projects = db.relationship(
         Project, lazy='dynamic',
         primaryjoin=db.and_(
             Profile.id == Project.profile_id, Project.parent_id == None,
-            db.or_(Project.state.DRAFT, Project.cfp_state.DRAFT)),
-        order_by=Project.date.desc())  # NOQA
+            db.or_(Project.state.DRAFT, Project.cfp_state.DRAFT)))  # NOQA
```
Comment (jace) [inline funnel/models/project.py]: We should start dropping `# NOQA` in query statements by switching to `column.is_(None)` and `column.isnot(None)`

File: funnel/templates/project.html.jinja2 [hunk 13]
```diff
@@ -14,7 +14,7 @@
     "@type": "Event",
     "name": "{{ project.title }}",
     "url": "{{ project.url_for(_external=true) }}",
-    "startDate": "{{ project.date }}",
+    "startDate": "{{ project.schedule_start_at.isoformat() }}",
```
Comment (jace) [inline funnel/templates/project.html.jinja2]: Does this need to be a date or a datetime?

File: funnel/views/schedule.py (hunk 8 above)
Comment (jace) [reply funnel/views/schedule.py]: Looks like they need to be cast into the project's timezone though.
